backend/src/apis/fetch.rs

```diff
@@ -1,151 +1,102 @@
-use base64::encode;
-use reqwest::header::{HeaderMap, HeaderValue, AUTHORIZATION, CONTENT_TYPE};
-use serde::Deserialize;
-use std::error::Error;
-
-#[derive(Deserialize, Debug)]
-/// This is the root level of the JSON response. It contains a data field, which corresponds to the "data" field in the JSON.
-struct Response {
-    data: Data,
-}
-/// This struct represents the "data" field of the JSON. It contains an items field, which corresponds to the "items" array in the JSON.
-#[derive(Deserialize, Debug)]
-struct Data {
-    items: Vec<Item>,
-}
-
-/// This struct represents each object in the "items" array. It contains a balance field and a contract_address field. These fields correspond to the "balance" and "contract_address" fields in each item of the "items" array in the JSON.
-#[derive(Deserialize, Debug)]
-struct Item {
-    balance: String,
-    contract_address: String,
-}
-
-/// This function takes a list of asset contracts addresses, a list of addresses and returns the aggregated balance of these address PER EACH asset
-pub fn fetch_asset_sums(
-    addresses: Vec<String>,
-    asset_contract_addresses: Vec<String>,
-) -> Result<Vec<u64>, Box<dyn Error>> {
-    // create asset sums vector
-    let mut asset_sums: Vec<u64> = Vec::new();
-
-    // for each address in addresses vector call fetch_balances_per_addr and increment the asset_sums vector
-    for address in addresses {
-        let balances = fetch_balances_per_addr(address.clone(), asset_contract_addresses.clone())?;
-        for (i, balance) in balances.iter().enumerate() {
-            if asset_sums.len() <= i {
-                asset_sums.push(*balance);
-            } else {
-                let sum = asset_sums[i] + balance;
-                asset_sums[i] = sum;
-            }
-        }
-    }
-
-    Ok(asset_sums)
+use futures::future::try_join_all;
+use std::{error::Error, sync::Arc};
+
+use ethers::{
+    abi::Address,
+    contract::builders::ContractCall,
+    prelude::SignerMiddleware,
+    providers::{Http, Middleware, Provider},
+    signers::LocalWallet,
+    types::{H160, U256},
+};
+
+use crate::contracts;
+use contracts::generated::mock_erc20::MockERC20;
+
+pub trait BalanceFromContract<M: Middleware> {
```

Comment:
What is this trait for?

Reply:
I believe this one is for retrieving the non-ETH balances. In this case, though, I'd prefer it to be called TokenBalance with a function "get_token_balance" because we're dealing with assets but not some arbitrary smart contracts.

Reply:
> What is this trait for?

It's a trait designed to wrap the `ERC` token contract interfaces (e.g., `MockERC20`) that are generated by `abigen`.

For instance, consider two token contracts, `MockERC20_A` and `MockERC20_B`. Each contract interface file has a balance_of method. However, the `fetch_asset_sums` function could not directly call the `balance_of` function without utilizing an enum pattern. But with this pattern, we would have to add both `MockERC20_A` and `MockERC20_B` to the enum. If the CEX supports additional tokens, the enum type would become increasingly lengthy.

Instead of using enum, I implemented the `BalanceFromContract` trait for the `contracts` inputs. While this approach still requires some code modifications, we can potentially reduce the task by using `abigen` to generate the necessary bindings (interface files).

> I believe this one is for retrieving the non-ETH balances. In this case, though, I'd prefer it to be called TokenBalance with a function "get_token_balance" because we're dealing with assets but not some arbitrary smart contracts.

 I appreciate your suggestions and thank you for helping with the naming!

Reply:
Ok thanks for clarifying it! I think it makes sense to keep the pattern as you suggested following the naming proposed by Alex!

Reply:
Changed names at 9ada33502519